src/dune_engine/build_system.ml

```diff
@@ -550,8 +550,18 @@ end = struct
           in
           let* () =
             maybe_async_rule_file_op (fun () ->
-              Path.Build.Set.iter targets.files ~f:Path.Build.unlink_no_err;
-              Path.Build.Set.iter targets.dirs ~f:(fun dir -> Path.rm_rf (Path.build dir)))
+              let remove_target_dir dir = Path.rm_rf (Path.build dir) in
+              let remove_target_file path =
+                match Path.Build.unlink_status path with
+                | Success -> ()
+                | Is_a_directory ->
+                  (* If target changed from a directory to a file, delete
+                     in anyway. *)
+                  remove_target_dir path
+                | Other_error -> ()
```

Comment:
Maybe worth leaving a trace in the log when hitting this case, mentioning the specific exception?

Reply:
yes, done!